### DISCUSSION TITLE
Creating React output from the HTML builder

### DISCUSSION THREAD

**Discussion by pcolmer** [sphinx-doc/sphinx#12648] [Q&A]
I'm trying to build a website, using React, that will act as an internal library of all documentation written with Sphinx. The idea is that each author will retain ownership and control of their git repositories but the build process, instead of emitting full HTML files, will instead emit React Typescript files that are then built as part of the website build process.

I've got a fairly simple React template and I've amended it to become the `layout.html` file:

```
import {
    AppLayout,
    Container,
    ContentLayout,
    Header,
    SideNavigation,
  } from "@cloudscape-design/components";
  import { useRouter } from "next/router";
  
  const DocsTemplate = () => {
    const router = useRouter();
    return (
      <AppLayout
        toolsHide
        content={
          <ContentLayout header={<Header>{{ title|striptags|e }}{{ titlesuffix }}</Header>}>
            <Container>{% block body %} {% endblock %}</Container>
          </ContentLayout>
        }
      />
    );
  };
  
  export default DocsTemplate;
```

`theme.conf`:
```
[theme]
inherit = none
stylesheet = none

[options]
file_suffix = "tsx"
```
then in `config.py`, I'm specifying the name of the HTML theme.

However, when I run `sphinx-build source build`, the output files still have `.html` as the file extension, and not `.tsx`. I'm not sure why this isn't working ...

Also, the images are put into "../_images" which may not work with the React site. Is there a configuration setting I can use to override that location?

**Comment by AA-Turner** [comment]
I would try ``html_file_suffix = '...'``

See: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_file_suffix

For your second point, see https://github.com/sphinx-doc/sphinx/issues/10906 -- help is welcomed!

A

**Reply by pcolmer** [reply]
@AA-Turner I think I've got a solution for #10906. Can you please clarify which branch I should use to base the code off? `v7.4.7` or `master`?

**Reply by AA-Turner** [reply]
`master` for new features, please.

A

**Reply by pcolmer** [reply]
@AA-Turner thank you. I _thought_ I had implemented a good, clean solution but, in trying out different builders like `JSONHTMLBuilder`, I'm hitting a problem with the way I've defined the additional configuration values. I had hoped/expected any builder that extended or used `HTMLBuilder` to benefit automatically from the configuration extension but that doesn't seem to be the case.

I'm wondering now if I should move the configuration values into the core Sphinx config rather than just have `HTMLBuilder` "add" them to the config since the latter seems to be a problem.

**Comment by pcolmer** [comment]
> I would try html_file_suffix = '...'

Unfortunately, that doesn't seem to have made any difference. Running `make clean` then `sphinx-build source build` still gives me a bunch of files with a `.html` extension. I'll start digging into the source code to see if I can figure out why.

I'll also take a look at whether or not I can implement a solution to #10906. No promises ... :)

**Comment by pcolmer** [comment]
> Unfortunately, that doesn't seem to have made any difference. Running make clean then sphinx-build source build still gives me a bunch of files with a .html extension. I'll start digging into the source code to see if I can figure out why.

OK - got to the bottom of this, eventually.

The setting needs to be in `conf.py` in the documentation source. I was trying to set it in `theme.toml` and/or `theme.conf` for the template itself.

Is there a way to set this in the template settings file?

**Reply by AA-Turner** [reply]
There isn't, no. The `[options]` table defines options within the theme that can be changed by users, and the settings in the `[theme]` table are fairly minimal ([ref](https://www.sphinx-doc.org/en/master/development/html_themes/index.html#theme-configuration-theme-toml)).

What might be useful though is making your theme into an extension ([ref](https://www.sphinx-doc.org/en/master/development/html_themes/index.html#distribute-your-theme-as-a-python-package)), which allows controlling other user settings. 

For inspiration you could look at the source code of the Alabaster or Furo themes.

A

**Comment by AA-Turner** [comment]
I haven't looked into any of these in great detail, but may be of use:

- https://github.com/kaymccormick/docutils-react
- https://medium.com/@mtiller/rendering-sphinx-documentation-with-react-95b785293a76

It may be easier in the long run to define a custom `ReactHTMLBuilder` which allows for deeper control of Sphinx's internals...

A

**Reply by pcolmer** [reply]
Thanks, @AA-Turner. I was _hoping_ that the bulk of the heavy lifting could be done by using the processing being done by the existing HTML Builder. I'm still trying to understand the code, though. My Python knowledge is struggling with that. Hence why I was trying templates/themes as a "shortcut" first.
